### PR TITLE
Support simple completions in non-interactive context.  Support keyword completions 

### PR DESCRIPTION
**Pull request by geraldus**
Related #1171

### CODE REVIEW

**Comment by geraldus** [comment]
I'm modifying REPL completion function right now and can push few commits here.

**Comment by gracjan** [comment]
Good stuff. Can you remove these things from `haskell-interactive-mode`?

**Comment by geraldus** [comment]
Ok it looks like to work properly.

@gracjan didn't catch your point about interactive mode, what should I remove? (:

File: haskell-completions.el [hunk 1]
```diff
@@ -220,36 +274,71 @@ result only if prefix length is not less than MINLEN."
...
-            (list beg end lst)))))))
+        (when (not (eql typ 'haskell-completions-general-prefix))
+          ;; do not complete things in comments
+          (if (cl-member
+               typ
+               '(haskell-completions-pragma-name-prefix
+                 haskell-completions-ghc-option-prefix
+                 haskell-completions-language-extension-prefix))
+              ;; provide simple completions
+              (haskell-completions--simple-completions prefix-data)
```
Comment (gracjan) [inline haskell-completions.el]: As I understand there will be two completions hooks installed at the same time, one by `haskell-mode` other by `interactive-haskell-mode`. The non-interactive one will return completions for PRAGMA, GHC_OPTION and LANGUAGE, so there is no need to return those completions here again.

Unless I'm mistaken how the mechanism works.

Comment (geraldus) [inline haskell-completions.el]: @gracjan I tested this, and if first function in hook returns `nil` there will no completions.  I tested this creating two dummy completion functions and added both of them to `completion-at-point-functions` hook in elisp mode.  All time only first function was used to return completion candidates.  I don't know why it works this way though.

**Comment by geraldus** [comment]
As for `haskell-interactive-mode-completion-at-point-function`, to remove it we have to support REPL command completions.  `haskell-completions-sync-repl-completions` can give right candidates if we take care of `:` when grabbing prefix in REPL buffer.

**Comment by fice-t** [comment]
Should the hardcoded completions have extra information for support for `eldoc-mode`? That is, in `haskell-doc-mode`,  [there are a few huge hardcoded alists](https://github.com/haskell/haskell-mode/blob/master/haskell-doc.el#L465) that are currently outdated.

I was thinking that you make new giant alists so that `eldoc` could use both keys and values, where the completion would just use the keys.

**Comment by geraldus** [comment]
@fice-t hmm, I didn't catch your idea, if we do have such list already why we need another one?

**Comment by gracjan** [comment]
I wonder how useful the hints by haskell-doc-mode are. Old timers know basic language constructs by heart, can newbies sort out keyword usage from the very short descriptions provided by haskell-doc-mode?

**Comment by geraldus** [comment]
I like doc messages for top-level forms, but keyword doc strings seem for me quite confusing :D

**Comment by geraldus** [comment]
By the way, @gracjan in Emacs 25 I have colored message in mini-buffer from `eldoc`, should we follow this (I remember in past you suggested not to fontify haskell-doc messages)?

**Comment by fice-t** [comment]
The lists are old as they're for Haskell-98. According to git they haven't been updated in over a decade.

**Comment by geraldus** [comment]
@fice-t I agree, I saw that before, I just do not understand what benefits of extra information in hard-coded lists?

**Comment by fice-t** [comment]
The extra information would be used in `eldoc` hints. For example:

`("filter" . "(a -> Bool) -> [a] -> [a]")`
The completion could just use `filter`, but `eldoc` could use both to create `filter :: (a-> Bool) -> [a] -> [a]`. The same for the keywords (though that might not be as useful).

So it's just using the same list for different purposes.

**Comment by geraldus** [comment]
@fice-t, pal, I'm still confused.  Look, `haskell-completions--pragma-names` and `haskell-completions--keywords` lists are used only to provide completion candidates, that is either when `complete-symbol` function is invoked, or completion functions from `haskell-completions.el`.  They have no relation to `haskell-doc.el`, it have separate hard-coded lists, and I see doc-strings in mini-buffer for keywords and top-level form when haskell session is present (and in my configuration `haskell-doc` mode is not enabled, `Global-Eldoc` somehow magically gets information from `haskell-doc` itself).  Do you facing some issues with `haskell-doc`?

File: haskell-completions.el (hunk 1 above)
Comment (gracjan) [inline haskell-completions.el]: I've repeated this experiment and it works like described in Emacs docs (and unlike you say). That means I'm able to complete "1" to "123" and "ab" to any of the options listed. Emacs seems to work correctly.

**Comment by gracjan** [comment]
I still think the static and the interactive completion functions should be completely separate.

For the interactive case there should be only `haskell-completions-grab-identifier-prefix` called, this should return `nil` if inside of a comment.

For the static case there should be only `haskell-completions-grab-pragma-prefix` called, this should return `nil` if outside of a pragma comment.

Emacs will call the hooks one by one, stopping with the one that returns non-nil value. This seems perfectly reasonable to me.

Am I missing something?

**Comment by geraldus** [comment]
@gracjan sounds reasonable, but
1. you forgot about keywords in static case

> Emacs will call the hooks one by one

2. can you please test this (as I mentioned before in my tests second completion function was not triggered in case when first returned `nil`, I updated my comment previous comment with code I've used to tests).

**Comment by fice-t** [comment]
>  Look, haskell-completions--pragma-names and haskell-completions--keywords lists are used only to provide completion candidates, that is either when complete-symbol function is invoked, or completion functions from haskell-completions.el.

That's why I'm saying this. I'm saying (I believe) there could be a single list which both completion and hints could use in some cases, such as keywords and basic/Prelude functions. That might be beneficial as one could maintain just a single list (and periodically generate a separate completion list if one was to speed completion up).

> Global-Eldoc somehow magically gets information from haskell-doc itself)

Yeah, `eldoc-documentation-function` is bound to a function that calls another function that checks the hardcoded `haskell-doc` lists.

**Comment by geraldus** [comment]
@fice-t ok, now I got it :D 
First of all, both lists are incomplete in our case according to wiki page, they do not include operator like things like `<-`, `->` and friends.  While it feel sensible to add this stuff to `haskell-doc.el`, this is not a case for completions.  Secondly I added to completion list special entries such `data family`, `import qualified`. Why?  For example type `im`, in case of `company-mode` this will give you both `import` and `import qualified` candidates in list, in case of vanilla Emacs hitting <kbd>TAB</kbd> or <kbd>C-M-i</kbd> first expand it to `import`, consequential command invocation will pop-up a Completion List buffer where both candidates are listed.  In we'll use keywords from `haskell-doc` we loose this feature, though it's possible to have a tiny list with such improved candidates and append it haskell-doc list.

**Comment by geraldus** [comment]
@gracjan this is my thought about your proposal.

1. I suppose that keywords should be completed in both cases
2. there is `:exclusive` property allowing to let Emacs run next completion function in case when it can't find matching candidate in list returned by current function.  Initially I though this could solve our puzzle, however, if static function will return keywords list with `:exclusive no` property user won't get other interactive completions if thing to be expanded matches some keyword, that is
`ca` with give sole completion `case`, but it is possible, that interactive function could return much more candidates, only when user will type `car` Emacs will call interactive completions.
3. if we follow you proposal we have to make sure that interactive completion function should be preceded by static, this will require first remove static function from hook and then add both in desired order.  Current implementation requires reversed order, e.g. interactive function should succeed static one and this is done out-of-box.

**Comment by gracjan** [comment]
@geraldus: I think you convinced me that Emacs default handling of `competion-at-point-functions` is not appropriate and haskell-mode should handle completions on its own.

In this case you should implement it as you have done. Ping me when the PR is ready to merge.

**Pull request merged**